# DATA UNDERSTANDING

In [385]:
##Import Relevant Libraries
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns 
%matplotlib inline

In [386]:
##Import dataset 
df = pd.read_csv(r'C:\Users\David\PHASE 3 PROJECT\PHASE-3-PROJECT\bigml_59c28831336c6604c800002a.csv')  
print(df)

     state  account length  area code phone number international plan  \
0       KS             128        415     382-4657                 no   
1       OH             107        415     371-7191                 no   
2       NJ             137        415     358-1921                 no   
3       OH              84        408     375-9999                yes   
4       OK              75        415     330-6626                yes   
...    ...             ...        ...          ...                ...   
3328    AZ             192        415     414-4276                 no   
3329    WV              68        415     370-3271                 no   
3330    RI              28        510     328-8230                 no   
3331    CT             184        510     364-6381                yes   
3332    TN              74        415     400-4344                 no   

     voice mail plan  number vmail messages  total day minutes  \
0                yes                     25              

# Exploratory data analysis

In [387]:
# view dimensions of dataset
df.shape

(3333, 21)

There are 3333 instances and 21 variables in the data set.

In [388]:
# preview the dataset
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [389]:
# Get the data types of all columns
print(df.dtypes)

state                      object
account length              int64
area code                   int64
phone number               object
international plan         object
voice mail plan            object
number vmail messages       int64
total day minutes         float64
total day calls             int64
total day charge          float64
total eve minutes         float64
total eve calls             int64
total eve charge          float64
total night minutes       float64
total night calls           int64
total night charge        float64
total intl minutes        float64
total intl calls            int64
total intl charge         float64
customer service calls      int64
churn                        bool
dtype: object


In [390]:
col_names = df.columns

col_names

Index(['state', 'account length', 'area code', 'phone number',
       'international plan', 'voice mail plan', 'number vmail messages',
       'total day minutes', 'total day calls', 'total day charge',
       'total eve minutes', 'total eve calls', 'total eve charge',
       'total night minutes', 'total night calls', 'total night charge',
       'total intl minutes', 'total intl calls', 'total intl charge',
       'customer service calls', 'churn'],
      dtype='object')

In [391]:
print(df.dtypes)

state                      object
account length              int64
area code                   int64
phone number               object
international plan         object
voice mail plan            object
number vmail messages       int64
total day minutes         float64
total day calls             int64
total day charge          float64
total eve minutes         float64
total eve calls             int64
total eve charge          float64
total night minutes       float64
total night calls           int64
total night charge        float64
total intl minutes        float64
total intl calls            int64
total intl charge         float64
customer service calls      int64
churn                        bool
dtype: object


In [392]:
#find numerical variables
numerical = [var for var in df.columns if df[var].dtype!='O']

print('There are {} numerical variables\n'.format(len(numerical)))

print('The numerical variables are :', numerical)

There are 17 numerical variables

The numerical variables are : ['account length', 'area code', 'number vmail messages', 'total day minutes', 'total day calls', 'total day charge', 'total eve minutes', 'total eve calls', 'total eve charge', 'total night minutes', 'total night calls', 'total night charge', 'total intl minutes', 'total intl calls', 'total intl charge', 'customer service calls', 'churn']


There are 17 numerical variables in the dataset

In [393]:
#find categorical variables

categorical = [var for var in df.columns if df[var].dtype=='O']

print('There are {} categorical variables\n'.format(len(categorical)))

print('The categorical variables are :', categorical)

There are 4 categorical variables

The categorical variables are : ['state', 'phone number', 'international plan', 'voice mail plan']


In [394]:
# view the categorical variables

df[categorical].head()

state phone number international plan voice mail plan
0    KS     382-4657                 no             yes
1    OH     371-7191                 no             yes
2    NJ     358-1921                 no              no
3    OH     375-9999                yes              no
4    OK     330-6626                yes              no

Summary of categorical variables

There are two categorical variables. These are given by state and phone number.

There are two binary categorical variables - international plan and voice mail plan.

voice mail plan is the target variable.

### Explore problems in categorical variables

In [395]:
# check missing values in categorical variables

df[categorical].isnull().sum()

state                 0
phone number          0
international plan    0
voice mail plan       0
dtype: int64

There are no missing values in categorical variable.

### Frequency counts of categorical variables

In [396]:
# view frequency of categorical variables

for var in categorical: 
    
    print(df[var].value_counts())

WV    106
MN     84
NY     83
AL     80
OR     78
OH     78
WI     78
WY     77
VA     77
CT     74
ID     73
MI     73
VT     73
UT     72
TX     72
IN     71
MD     70
KS     70
NJ     68
MT     68
NC     68
CO     66
NV     66
WA     66
MA     65
MS     65
RI     65
AZ     64
MO     63
FL     63
NM     62
ME     62
ND     62
DE     61
NE     61
OK     61
SD     60
SC     60
KY     59
IL     58
NH     56
AR     55
GA     54
DC     54
HI     53
TN     53
AK     52
LA     51
PA     45
IA     44
CA     34
Name: state, dtype: int64
393-7984    1
369-7295    1
348-8711    1
420-5645    1
329-6562    1
           ..
371-4306    1
388-4879    1
370-8676    1
340-2239    1
376-4519    1
Name: phone number, Length: 3333, dtype: int64
no     3010
yes     323
Name: international plan, dtype: int64
no     2411
yes     922
Name: voice mail plan, dtype: int64


In [397]:
# view frequency distribution of categorical variables

for var in categorical: 
    
    print(df[var].value_counts()/np.float64(len(df)))

WV    0.031803
MN    0.025203
NY    0.024902
AL    0.024002
OR    0.023402
OH    0.023402
WI    0.023402
WY    0.023102
VA    0.023102
CT    0.022202
ID    0.021902
MI    0.021902
VT    0.021902
UT    0.021602
TX    0.021602
IN    0.021302
MD    0.021002
KS    0.021002
NJ    0.020402
MT    0.020402
NC    0.020402
CO    0.019802
NV    0.019802
WA    0.019802
MA    0.019502
MS    0.019502
RI    0.019502
AZ    0.019202
MO    0.018902
FL    0.018902
NM    0.018602
ME    0.018602
ND    0.018602
DE    0.018302
NE    0.018302
OK    0.018302
SD    0.018002
SC    0.018002
KY    0.017702
IL    0.017402
NH    0.016802
AR    0.016502
GA    0.016202
DC    0.016202
HI    0.015902
TN    0.015902
AK    0.015602
LA    0.015302
PA    0.013501
IA    0.013201
CA    0.010201
Name: state, dtype: float64
393-7984    0.0003
369-7295    0.0003
348-8711    0.0003
420-5645    0.0003
329-6562    0.0003
             ...  
371-4306    0.0003
388-4879    0.0003
370-8676    0.0003
340-2239    0.0003
376-4519    0.000

### Number of labels: cardinality

High cardinality may pose some serious problems in the machine learning model. So, I will check for high cardinality.

In [398]:
# check for cardinality in categorical variables

for var in categorical:
    
    print(var, ' contains ', len(df[var].unique()), ' labels')

state  contains  51  labels
phone number  contains  3333  labels
international plan  contains  2  labels
voice mail plan  contains  2  labels


We can see that there is a phone number variable which needs to be preprocessed. If a phone number field has the highest cardinality (i.e., each phone number is unique or nearly unique), it presents challenges for preprocessing, especially for machine learning models where high cardinality features can lead to inefficiencies and overfitting

All the other variables contain relatively smaller number of variables.

### Feature Engineering of phone number Variable

### Standardizing Phone Number

In [399]:
#standardizing phone numbers to a consistent format helps in comparison and analysis. 
#I will use libraries like phonenumbers for this purpose.

import phonenumbers
import pandas as pd


data = {
    'phone_number': ['382-4657', '371-7191	', '358-1921', '375-9999','330-6626	']
}
df = pd.DataFrame(data)

def standardize_phone_number(phone_number):
    try:
        parsed_number = phonenumbers.parse(phone_number, None)
        return phonenumbers.format_number(parsed_number, phonenumbers.PhoneNumberFormat.E164)
    except phonenumbers.NumberParseException:
        return None

df['standardized_phone_number'] = df['phone_number'].apply(standardize_phone_number)

In [400]:
#1.Extract Area Code or Country Code
def extract_country_code(phone_number):
    try:
        parsed_number = phonenumbers.parse(phone_number, None)
        return parsed_number.country_code
    except phonenumbers.NumberParseException:
        return None

df['country_code'] = df['standardized_phone_number'].apply(extract_country_code)


In [401]:
#2. Extract the Number Length
df['number_length'] = df['standardized_phone_number'].str.len()
df.head()

phone_number standardized_phone_number country_code number_length
0     382-4657                      None         None          None
1   371-7191\t                      None         None          None
2     358-1921                      None         None          None
3     375-9999                      None         None          None
4   330-6626\t                      None         None          None

In [402]:
def clean_phone_number(phone_number):
    cleaned_number = re.sub(r'\D', '', phone_number)
    return cleaned_number

# Apply the function again
df['cleaned_phone_number'] = df['phone_number'].apply(clean_phone_number)


In [403]:
# again view the summary of dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   phone_number               5 non-null      object
 1   standardized_phone_number  0 non-null      object
 2   country_code               0 non-null      object
 3   number_length              0 non-null      object
 4   cleaned_phone_number       5 non-null      object
dtypes: object(5)
memory usage: 328.0+ bytes


There are 4 additional columns added to the phone number. These are the Cleaned Phone Number,country code, number length and the standardized phone number .I will drop the original phone number.

In [404]:
# drop the original phone number variable

df.drop('phone_number', axis=1, inplace = True)

In [405]:
# preview the dataset again

df.head()

standardized_phone_number country_code number_length cleaned_phone_number
0                      None         None          None              3824657
1                      None         None          None              3717191
2                      None         None          None              3581921
3                      None         None          None              3759999
4                      None         None          None              3306626

Now, we can see that the phone number variable has been removed from the dataset.

### Explore Categorical Variables

I will explore the categorical variables one by one.

In [406]:
# find categorical variables

categorical = [var for var in df.columns if df[var].dtype=='O']

print('There are {} categorical variables\n'.format(len(categorical)))

print('The categorical variables are :', categorical)

There are 4 categorical variables

The categorical variables are : ['standardized_phone_number', 'country_code', 'number_length', 'cleaned_phone_number']


In [407]:
# view the categorical variables

df[categorical].head()

standardized_phone_number country_code number_length cleaned_phone_number
0                      None         None          None              3824657
1                      None         None          None              3717191
2                      None         None          None              3581921
3                      None         None          None              3759999
4                      None         None          None              3306626

In [408]:
# check for missing values in categorical variables 

df[categorical].isnull().sum()



standardized_phone_number    5
country_code                 5
number_length                5
cleaned_phone_number         0
dtype: int64

We can see that cleaned phone number has no missing values while standardadized phone number, country code and number length have 5 missing values each.

### Explore cleaned_phone_number Variable

In [409]:
# print number of labels in cleaned_phone_number  variable

print('cleaned_phone_number  contains', len(df["standardized_phone_number"].unique()), 'labels')

cleaned_phone_number  contains 1 labels


In [410]:
# check labels in cleaned_phone_number variable

df.cleaned_phone_number.unique()

array(['3824657', '3717191', '3581921', '3759999', '3306626'],
      dtype=object)

In [411]:
# check frequency distribution of values in cleaned_phone_number variable

df["cleaned_phone_number"].value_counts()

3306626    1
3759999    1
3717191    1
3581921    1
3824657    1
Name: cleaned_phone_number, dtype: int64

In [412]:
# let's do One Hot Encoding of cleaned_phone_number variable
# get k-1 dummy variables after One Hot Encoding 
# preview the dataset with head() method

pd.get_dummies(df["cleaned_phone_number"], drop_first=True).head()

3581921  3717191  3759999  3824657
0        0        0        0        1
1        0        1        0        0
2        1        0        0        0
3        0        0        1        0
4        0        0        0        0

### Explore standardized_phone_number  variable

In [413]:
# print number of labels in standardized_phone_number  variable

print('standardized_phone_number  contains', len(df["standardized_phone_number"].unique()), 'labels')

standardized_phone_number  contains 1 labels


In [414]:
# check labels in standardized_phone_number variable

df.standardized_phone_number.unique()

array([None], dtype=object)

In [415]:
# let's do One Hot Encoding of WindDir3pm variable
# get k-1 dummy variables after One Hot Encoding 
# also add an additional dummy variable to indicate there was missing data
# preview the dataset with head() method

pd.get_dummies(df['standardized_phone_number'], drop_first=True, dummy_na=True).head()

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4]

### Explore country_code variable

In [416]:
# print number of labels in country_code variable

print('country_code  contains', len(df["country_code"].unique()), 'labels')

country_code  contains 1 labels


In [417]:
#check labels in country_code variable
df.country_code.unique()

array([None], dtype=object)

In [418]:
# check frequency distribution of values in country_code variable

df["country_code"].value_counts()

Series([], Name: country_code, dtype: int64)

In [419]:
# let's do One Hot Encoding of WindDir3pm variable
# get k-1 dummy variables after One Hot Encoding 
# also add an additional dummy variable to indicate there was missing data
# preview the dataset with head() method

pd.get_dummies(df["country_code"], drop_first=True, dummy_na=True,dtype="int").head()

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4]

### DATA VISUALIZATION

In [421]:
plt.figure(figsize=(5, 6))
sns.scatterplot(x='total_day_minutes', y='total_eve_minutes', hue='international_plan', data=df)
plt.show()
plt.figure(figsize=(5, 6))
sns.boxplot(x='voice_mail_plan', y='total_eve_minutes', data=df)
plt.show()

ValueError: Could not interpret value `total_day_minutes` for parameter `x`

<Figure size 360x432 with 0 Axes>

## Declare feature vector and target variable

In [ ]:
X = df.drop(['cleaned_phone_number'], axis=1)

y = df['cleaned_phone_number']


## Split data into separate training and test set

In [ ]:
# split X and y into training and testing sets

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [ ]:
# check the shape of X_train and X_test

X_train.shape, X_test.shape
((113754, 24), (28439, 24))

((113754, 24), (28439, 24))

## Feature Engineering

Here I will transform raw data into useful features that help us to understand our model better and increase its predictive power. I will carry out feature engineering on the categorical variable.

In [ ]:
# check data types in X_train

X_train.dtypes

standardized_phone_number    object
country_code                 object
number_length                object
dtype: object

In [ ]:
# display categorical variables

categorical = [col for col in X_train.columns if X_train[col].dtypes == 'O']

categorical

['standardized_phone_number', 'country_code', 'number_length']

In [ ]:
# display numerical variables

numerical = [col for col in X_train.columns if X_train[col].dtypes != 'O']

numerical

[]

### Engineering missing values in categorical variables

In [ ]:
# print percentage of missing values in the categorical variables in training set

X_train[categorical].isnull().mean()

standardized_phone_number    1.0
country_code                 1.0
number_length                1.0
dtype: float64

In [ ]:
# print categorical variables with missing data

for col in categorical:
    if X_train[col].isnull().mean()>0:
        print(col, (X_train[col].isnull().mean()))

standardized_phone_number 1.0
country_code 1.0
number_length 1.0


In [ ]:
X_train.mode()

Empty DataFrame
Columns: [standardized_phone_number, country_code, number_length]
Index: []

In [ ]:
# check missing values in categorical variables in X_train

X_train[categorical].isnull().sum()

standardized_phone_number    4
country_code                 4
number_length                4
dtype: int64

In [ ]:
#confirm the columns that we are working with
print(df.columns)

Index(['standardized_phone_number', 'country_code', 'number_length',
       'cleaned_phone_number'],
      dtype='object')


In [ ]:
# impute missing categorical variables with most frequent value
# Fill missing categorical values with the mode
df['standardized_phone_number'] = df['standardized_phone_number'].astype('category')
df['standardized_phone_number'] = df['standardized_phone_number'].cat.add_categories('Missing').fillna('Missing')



In [ ]:
# check missing values in categorical variables in X_train

X_train[categorical].isnull().sum()

standardized_phone_number    4
country_code                 4
number_length                4
dtype: int64

In [ ]:
# check missing values in categorical variables in X_test

X_test[categorical].isnull().sum()

standardized_phone_number    1
country_code                 1
number_length                1
dtype: int64

As a final check, I will check for missing values in X_train and X_test.

In [ ]:
# check missing values in X_train

X_train.isnull().sum()

standardized_phone_number    4
country_code                 4
number_length                4
dtype: int64

In [ ]:
# check missing values in X_test

X_test.isnull().sum()

standardized_phone_number    1
country_code                 1
number_length                1
dtype: int64

We can see that there are no missing values in X_train and X_test

### Encode categorical variables

In [ ]:
categorical

['standardized_phone_number', 'country_code', 'number_length']

In [ ]:
X_train[categorical].head()

standardized_phone_number  country_code  number_length
0                        NaN           NaN            NaN
1                        NaN           NaN            NaN
3                        NaN           NaN            NaN
4                        NaN           NaN            NaN

In [ ]:
# encode standardized_phone_number variable
import category_encoders as ce

binary_encoder = ce.BinaryEncoder(cols=['standardized_phone_number', 'country_code', 'number_length'])

X_train = binary_encoder.fit_transform(X_train)

X_test = binary_encoder.transform(X_test)

X_train.head()

standardized_phone_number_0  country_code_0  number_length_0
0                            1               1                1
1                            1               1                1
3                            1               1                1
4                            1               1                1

In [ ]:
X_train.isnull().sum()

standardized_phone_number_0    0
dtype: int64

We can see there is an addidtional column added standardized phone number_0

Now, I will create the X_train training set.